src/crimson/tools/store_bench/store-bench.cc

```diff
@@ -0,0 +1,685 @@
+
+#include "crimson/common/config_proxy.h"
+#include "crimson/common/coroutine.h"
+#include "crimson/common/log.h"
+
+#include "crimson/os/futurized_collection.h"
+#include "crimson/os/futurized_store.h"
+
+namespace po = boost::program_options;
+
+using namespace ceph;
+
+SET_SUBSYS(osd);
+
+/**
+ *These are functions that are used in both types of work_loads
+ */
+
+/**
+ * Creates a unique object
+ */
+ghobject_t create_hobj(unsigned id) {
+  return ghobject_t(
+      shard_id_t::NO_SHARD, seastar::this_shard_id(),
+      id, // hash, normally rjenkins of name, but let's just set it to id
+      "", // namespace, empty here
+      "", // name, empty here
+      0,  // snapshot
+      ghobject_t::NO_GEN);
+};
+
+/**
+ * This function creates a collection(PG) with pool id as obj_id
+ * The function is called when we create an object
+ * the goal is to have each object belong to a separate collection
+ */
+coll_t make_cid(int obj_id) { return coll_t(spg_t(pg_t(obj_id, 0))); };
+
+/**
+ * The struct stores the results for a single io
+ * The number of operations is incrmented by 1 for write+remove i =n the case of
+ * pg_workload for rgw workload, each write increments operations by 1 , each
+ * delte increments op by 1 tot_latency per_io is the sum of times it takes for
+ * all the operations per io
+ */
+struct results_t {
+  int num_operations;
+  int tot_latency_per_io;
+};
+
+/**
+ * This function is used to run a workload num_con_io times in parallel
+ * the  parallel_for_each loop launches k co_routines in parallel
+ * Each coroutine is the actual work_load function which returns a
+ *future<result> Since its a co_await once that future is resolved it is added
+ *to the vector so all_io_res containes resolved futures
+ */
+seastar::future<>
+run_concurrent_ios(int duration, int num_concurrent_io,
+                   std::function<seastar::future<results_t>()> work_load) {
+  LOG_PREFIX(run_concurrent_ios);
+  std::vector<int> container_io;
+  std::vector<results_t> all_io_res;
+  for (int i = 0; i < num_concurrent_io; ++i) {
+    container_io.push_back(i);
+  }
+  co_await seastar::parallel_for_each(
+      container_io, (seastar::coroutine::lambda([&](int) -> seastar::future<> {
+        auto res = co_await work_load();
+        all_io_res.push_back(std::move(res));
+        co_return;
+      })));
+
+  int tot_ops_all_io = 0;
+  int tot_latency_all_io = 0;
+  for (auto it = all_io_res.begin(); it != all_io_res.end(); ++it) {
+    tot_ops_all_io += it->num_operations;
+    tot_latency_all_io += it->tot_latency_per_io;
+  }
+  ERROR("Total number of operations performed across ios is {}",
+        tot_ops_all_io);
+  ERROR("Total latency aka time per operation, across ios is {}",
+        tot_latency_all_io);
+  ERROR("throughput, number of ops per millisec is {}",
+        static_cast<double>(tot_ops_all_io) / duration);
+  ERROR("average latency per io is  {}",
+        tot_latency_all_io / tot_ops_all_io);
+  co_return;
```

Comment:
This could return a results_t allowing the caller to decide how to report the results.